napari/util/keybindings.py

```diff
@@ -0,0 +1,372 @@
+    but can be denoted to work on release too by separating the function
+    into two statements with the yield keyword::
+
+        @viewer.bind_key('h')
+        def hello_world(viewer):
+            # on key press
+            viewer.status = 'hello world!'
+
+            yield
+
+            # on key release
+            viewer.status = 'goodbye world :('
+
+    """
+    if func is UNDEFINED:
+
+        def inner(func):
+            bind_key(keymap, key, func, overwrite=overwrite)
+            return func
+
+        return inner
+
+    key = normalize_key_combination(key)
+
+    if func is not None and key in keymap and not overwrite:
+        raise ValueError(
+            f'key combination {key} already used! '
+            "specify 'overwrite=True' to bypass this check"
+        )
+
+    unbound = keymap.pop(key, None)
+
+    if func is not None:
+        if not callable(func):
+            raise TypeError("'func' must be a callable")
+        keymap[key] = func
+
+    return unbound
+
+
+class KeybindingDescriptor:
+    """Descriptor which transforms ``func`` into a method with the first argument bound
+    to ``class_keymap`` or ``_keymap`` depending on if it was called
+    from the class or the instance, respectively.
+
+    Parameters
+    ----------
+    func : callable
+        Function to bind.
+    """
+
+    def __init__(self, func):
+        self.__func__ = func
+
+    def __get__(self, instance, klass):
+        if instance is None:  # used on class
+            try:
+                keymap = klass.class_keymap
+            except AttributeError:
+                return self.__func__
+        else:
+            keymap = instance._keymap
+
+        return types.MethodType(self.__func__, keymap)
+
+
+class InheritedKeymap(UserDict):
+    """Dictionary which inherits from another.
+
+    Values of ``None`` are treated as though the key doesn't exist.
+
+    Parameters
+    ----------
+    parent : callable() -> dict
+        Parent keymap.
+    contents : dict, optional
+        Contents with which to initialize.
+    """
+
+    def __init__(self, parent, contents={}):
+        super().__init__(contents)
+        self._parent = parent
+
+    def __getitem__(self, name):
+        try:
+            val = super().__getitem__(name)
+        except KeyError:
+            val = self._parent()[name]
+
+        if val is None:
+            raise KeyError(name)
+
+        return val
+
+    def __delitem__(self, name):
+        self[name] = None
+
+    def copy(self):
+        return InheritedKeymap(self._parent, self.data.copy())
+
+    def pop(self, name, default=UNDEFINED):
+        try:
+            default = self[name]
+        except KeyError:
+            if default is UNDEFINED:
+                raise
+        else:
+            del self[name]
+
+        return default
+
+
+class KeymapMixin:
+    """Mix-in to add keymap functionality. Must still define ``class_keymap``
+    in every subclass.
```

Comment:
Is there really no way around this?

Reply:
Reiterating for those not following along the full conversation: we could use a metaclass which would be pretty gross too.